File: app/src/main/java/org/stepic/droid/ui/activities/SplashActivity.java
package org.stepic.droid.ui.activities;


import android.content.Intent;
import android.content.pm.ShortcutInfo;
import android.content.pm.ShortcutManager;
import android.graphics.drawable.Icon;
import android.os.Bundle;

import org.stepic.droid.R;
import org.stepic.droid.analytic.Analytic;
import org.stepic.droid.base.App;
import org.stepic.droid.core.presenters.SplashPresenter;
import org.stepic.droid.core.presenters.contracts.SplashView;
import org.stepic.droid.notifications.StepicInstanceIdService;
import org.stepic.droid.util.AppConstants;

import java.util.Arrays;

import javax.inject.Inject;

import kotlin.Unit;
import kotlin.jvm.functions.Function0;


public class SplashActivity extends BackToExitActivityBase implements SplashView {

    @Inject
    SplashPresenter splashPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //This stops from opening again from the Splash screen when minimized
        if (!isTaskRoot()) {
            finish();
            return;
        }
        App.Companion.componentManager().splashComponent().inject(this);
        splashPresenter.attachView(this);
        splashPresenter.onSplashCreated();

        defineShortcuts();


        if (checkPlayServices() && !sharedPreferenceHelper.isGcmTokenOk()) {
            threadPoolExecutor.execute(new Runnable() {
                @Override
                public void run() {
                    StepicInstanceIdService.Companion.updateAnywhere(api, sharedPreferenceHelper, analytic); //FCM!
                }
            });
        }

        if (savedInstanceState == null) {
            threadPoolExecutor.execute(new Runnable() {
                @Override
                public void run() {
                    int numberOfLaunches = sharedPreferenceHelper.incrementNumberOfLaunches();
                    //after first increment it is 0, because of default value is -1.
                    if (numberOfLaunches <= 0) {
                        analytic.reportEvent(Analytic.System.FIRST_LAUNCH_AFTER_INSTALL);
                    }
                    if (numberOfLaunches < AppConstants.LAUNCHES_FOR_EXPERT_USER) {
                        analytic.reportEvent(Analytic.Interaction.START_SPLASH, numberOfLaunches + "");
                    } else {
                        analytic.reportEvent(Analytic.Interaction.START_SPLASH_EXPERT, numberOfLaunches + "");
                    }
                }
            });
        }

        if (sharedPreferenceHelper.isFirstTime() || !sharedPreferenceHelper.isScheduleAdded() || sharedPreferenceHelper.isNeedDropCoursesIn114()) {
            //fix v11 bug:
            threadPoolExecutor.execute(new Runnable() {
                @Override
                public void run() {
                    if (sharedPreferenceHelper.isFirstTime()) {
                        databaseFacade.dropOnlyCourseTable(); //v11 bug, when slug was not cached. We can remove it, when all users will have v1.11 or above. (flavour problem)
                        sharedPreferenceHelper.afterFirstTime();
                        sharedPreferenceHelper.afterScheduleAdded();
                        sharedPreferenceHelper.afterNeedDropCoursesIn114();
                    } else if (!sharedPreferenceHelper.isScheduleAdded()) {
                        databaseFacade.dropOnlyCourseTable();
                        sharedPreferenceHelper.afterScheduleAdded();
                        sharedPreferenceHelper.afterNeedDropCoursesIn114();
                    } else if (sharedPreferenceHelper.isNeedDropCoursesIn114()) {
                        databaseFacade.dropOnlyCourseTable();
                        sharedPreferenceHelper.afterNeedDropCoursesIn114();
                    }

                    mainHandler.post(new Function0<Unit>() {
                        @Override
                        public Unit invoke() {
                            checkRemoteConfigs();
                            return Unit.INSTANCE;
                        }
                    });

                }
            });

        } else {
            checkRemoteConfigs();
        }

    }

    private void defineShortcuts() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N_MR1) {
            ShortcutManager shortcutManager = getSystemService(ShortcutManager.class);

            String catalogLabel = getString(R.string.catalog_title);
            Intent catalogIntent = screenManager.getCatalogIntent(getApplicationContext());
            catalogIntent.setAction(AppConstants.OPEN_SHORTCUT_CATALOG);
            ShortcutInfo catalogShortcut = new ShortcutInfo.Builder(this, AppConstants.CATALOG_SHORTCUT_ID)
                    .setShortLabel(catalogLabel)
                    .setLongLabel(catalogLabel)
                    .setIcon(Icon.createWithResource(this, R.drawable.ic_shortcut_find_courses))
                    .setIntent(catalogIntent)
                    .build();

            String profileLabel = getString(R.string.profile_title);
            Intent profileIntent = screenManager.getMyProfileIntent(getApplicationContext());
            profileIntent.setAction(AppConstants.OPEN_SHORTCUT_PROFILE);
            ShortcutInfo profileShortcut = new ShortcutInfo.Builder(this, AppConstants.PROFILE_SHORTCUT_ID)
                    .setShortLabel(profileLabel)
                    .setLongLabel(profileLabel)
                    .setIcon(Icon.createWithResource(this, R.drawable.ic_shortcut_profile))
                    .setIntent(profileIntent)
                    .build();


            shortcutManager.setDynamicShortcuts(Arrays.asList(catalogShortcut, profileShortcut));
        }
    }

    private void checkRemoteConfigs() {
        //remove
    }

    private void showNextScreen() {
        if (!isFinishing()) {
            if (sharedPreferenceHelper.getAuthResponseFromStore() != null) {
                screenManager.showMainFeedFromSplash(SplashActivity.this);
            } else {
                screenManager.showLaunchFromSplash(this);
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        splashPresenter.detachView(this);
        if (isFinishing()) {
            App.Companion.componentManager().releaseSplashComponent();
        }
    }

    @Override
    public void onShowLaunch() {
        if (!isFinishing()) {
            screenManager.showLaunchFromSplash(this);
        }
    }

    @Override
    public void onShowHome() {
        if (!isFinishing()) {
            screenManager.showMainFeedFromSplash(SplashActivity.this);
        }
    }
}

File: app/src/main/java/org/stepic/droid/ui/activities/SplashActivity.kt
package org.stepic.droid.ui.activities


import android.annotation.TargetApi
import android.content.Intent
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.content.res.Resources
import android.graphics.drawable.Icon
import android.os.Bundle
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import io.branch.referral.Branch
import org.stepic.droid.R
import org.stepic.droid.analytic.AmplitudeAnalytic
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.analytic.BranchParams
import org.stepic.droid.analytic.experiments.OnboardingSplitTestVersion2
import org.stepic.droid.base.App
import org.stepic.droid.configuration.RemoteConfig
import org.stepic.droid.core.presenters.SplashPresenter
import org.stepic.droid.core.presenters.contracts.SplashView
import org.stepic.droid.util.AppConstants
import org.stepic.droid.util.defaultLocale
import org.stepik.android.domain.base.analytic.BUNDLEABLE_ANALYTIC_EVENT
import org.stepik.android.domain.base.analytic.toGenericAnalyticEvent
import org.stepik.android.view.routing.deeplink.BranchRoute
import java.util.Arrays
import javax.inject.Inject

class SplashActivity : BackToExitActivityBase(), SplashView {

    companion object {
        private const val RUSSIAN_LANGUAGE_CODE = "ru"
    }

    @Inject
    internal lateinit var splashPresenter: SplashPresenter

    @Inject
    internal lateinit var analytics: Analytic

    @Inject
    internal lateinit var firebaseRemoteConfig: FirebaseRemoteConfig

    @Inject
    internal lateinit var onboardingSplitTestVersion2: OnboardingSplitTestVersion2

    private val locale = Resources.getSystem().configuration.defaultLocale

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.componentManager().splashComponent().inject(this)
        splashPresenter.attachView(this)
//        if (!isTaskRoot) {
//            finish()
//            return
//        }
        if (savedInstanceState == null) {
            val analyticEvent = intent
                .getBundleExtra(BUNDLEABLE_ANALYTIC_EVENT)
                ?.toGenericAnalyticEvent()

            if (analyticEvent != null) {
                analytics.report(analyticEvent)
            }
        }
        defineShortcuts()
    }

    private fun defineShortcuts() {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.N_MR1) {
            return
        }
        val shortcutManager = getSystemService(ShortcutManager::class.java) ?: return

        val catalogIntent = screenManager.getCatalogIntent(applicationContext)
        catalogIntent.action = AppConstants.OPEN_SHORTCUT_CATALOG
        val catalogShortcut = createShortcutInfo(AppConstants.CATALOG_SHORTCUT_ID, R.string.catalog_title, catalogIntent, R.drawable.ic_shortcut_find_courses)

        val profileIntent = screenManager.getMyProfileIntent(applicationContext)
        profileIntent.action = AppConstants.OPEN_SHORTCUT_PROFILE
        val profileShortcut = createShortcutInfo(AppConstants.PROFILE_SHORTCUT_ID, R.string.profile_title, profileIntent, R.drawable.ic_shortcut_profile)

        shortcutManager.dynamicShortcuts = Arrays.asList(catalogShortcut, profileShortcut)
    }

    @TargetApi(25)
    private fun createShortcutInfo(id: String, @StringRes titleRes: Int, catalogIntent: Intent, @DrawableRes iconRes: Int): ShortcutInfo {
        val title = getString(titleRes)
        return ShortcutInfo.Builder(this, id)
                .setShortLabel(title)
                .setLongLabel(title)
                .setIcon(Icon.createWithResource(this, iconRes))
                .setIntent(catalogIntent)
                .build()
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        this.intent = intent
    }

    override fun onStart() {
        super.onStart()

        val uri = intent?.data
        if (uri == null) {
            splashPresenter.onSplashCreated()
        } else {
            Branch
                .sessionBuilder(this)
                .withCallback { referringParams, error ->
                    if (error == null && referringParams != null && referringParams.has(BranchParams.FIELD_CAMPAIGN)) {
                        analytics.reportAmplitudeEvent(AmplitudeAnalytic.Branch.LINK_OPENED, mapOf(
                            AmplitudeAnalytic.Branch.PARAM_CAMPAIGN to referringParams[BranchParams.FIELD_CAMPAIGN],
                            AmplitudeAnalytic.Branch.IS_FIRST_SESSION to referringParams.optBoolean(BranchParams.IS_FIRST_SESSION, false)
                        ))
                    }

                    splashPresenter.onSplashCreated(referringParams)
                }
                .withData(uri)
                .init()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        splashPresenter.detachView(this)
        if (isFinishing) {
            App.componentManager().releaseSplashComponent()
        }
    }

    override fun onShowLaunch() {
        screenManager.showLaunchFromSplash(this)
        finish()
    }

    override fun onShowHome() {
        screenManager.showMainFeedFromSplash(this)
        finish()
    }

    override fun onShowCatalog() {
        screenManager.showMainFeed(this, MainFeedActivity.CATALOG_INDEX)
    }

    override fun onShowOnboarding() {
        if (locale.language == RUSSIAN_LANGUAGE_CODE && firebaseRemoteConfig.getString(RemoteConfig.PERSONALIZED_ONBOARDING_COURSE_LISTS).isNotEmpty()) {
            when (onboardingSplitTestVersion2.currentGroup) {
                OnboardingSplitTestVersion2.Group.Control ->
                    screenManager.showOnboarding(this)
                OnboardingSplitTestVersion2.Group.Personalized ->
                    screenManager.showLaunchScreen(this)
                OnboardingSplitTestVersion2.Group.None ->
                    screenManager.showLaunchFromSplash(this)
                OnboardingSplitTestVersion2.Group.ControlPersonalized ->
                    screenManager.showOnboarding(this)
            }
        } else {
            sharedPreferenceHelper.setPersonalizedOnboardingWasShown()
            screenManager.showOnboarding(this)
        }
        finish()
    }

    override fun onDeepLinkRoute(route: BranchRoute) {
        screenManager.openDeepLink(this, route)
        finish()
    }
}
